k-distribution/src/test/scala/org/kframework/backend/kore/KoreTest.scala

```diff
@@ -0,0 +1,98 @@
+package org.kframework.backend.kore
+
+import org.kframework.kompile.Kompile
+import org.kframework.kompile.KompileOptions
+import org.kframework.main.GlobalOptions
+import org.kframework.utils.errorsystem.KExceptionManager
+import org.kframework.utils.file.FileUtil
+import org.kframework.utils.options.OuterParsingOptions
+import org.kframework.parser.kore._
+import org.kframework.parser.kore.implementation.{DefaultBuilders => B}
+import org.kframework.parser.kore.parser.TextToKore
+
+import java.io.File
+import java.nio.file.Files
+
+class KoreTest {
+
+  val global: GlobalOptions = new GlobalOptions()
+
+  val files: FileUtil = {
+    val tempRoot = Files.createTempDirectory("kore-test").toFile
+    val tempDir = new File(tempRoot, "tmp")
+    tempDir.mkdirs()
+    val kompiledDir = new File(tempRoot, "kompiled")
+    kompiledDir.mkdirs()
+    new FileUtil(tempDir, tempRoot, tempRoot, kompiledDir, global, System.getenv)
+  }
+
+  val kem: KExceptionManager = new KExceptionManager(global)
+
+  val options: KompileOptions = {
+    val result = new KompileOptions()
+    result.outerParsing = new OuterParsingOptions(files.resolveDefinitionDirectory("test.k"))
+    result
+  }
+
+  def kompile(k: String): Definition = {
+    val compiler = new Kompile(options, files, kem, false)
+    val backend = new KoreBackend(options, files, kem)
+    files.saveToDefinitionDirectory("test.k", k)
+    val defn = compiler.run(files.resolveDefinitionDirectory("test.k"), "TEST", "TEST", backend.steps, backend.excludedModuleTags)
+    backend.accept(defn)
+    new TextToKore().parse(files.resolveDefinitionDirectory("test.kore"))
+  }
+
+  def axioms(defn: Definition): Seq[AxiomDeclaration] = {
+    defn.modules.flatMap(_.decls.filter(_.isInstanceOf[AxiomDeclaration]).map(_.asInstanceOf[AxiomDeclaration]))
+  }
+
+  def axioms(k: String): Seq[AxiomDeclaration] = axioms(kompile(k))
+
+  def getRewrite(axiom: AxiomDeclaration): Option[GeneralizedRewrite] = {
+    def go(pattern: Pattern): Option[GeneralizedRewrite] = {
+      pattern match {
+        case And(_, Equals(_, _, _, _), And(_, _, rw @ Rewrites(_, _, _))) => Some(rw)
+        case And(_, Top(_), And(_, _, rw @ Rewrites(_, _, _))) => Some(rw)
+        case Rewrites(s, And(_, Equals(_, _, _, _), l), And(_, _, r)) => Some(B.Rewrites(s, l, r))
+        case Rewrites(s, And(_, Top(_), l), And(_, _, r)) => Some(B.Rewrites(s, l, r))
+        case Implies(_, Bottom(_), p) => go(p)
+        case Implies(_, Equals(_, _, _, _), And(_, eq @ Equals(_, _, Application(_, _), _), _)) => Some(eq)
+        case Implies(_, Top(_), And(_, eq @ Equals(_, _, Application(_, _), _), _)) => Some(eq)
+        case Implies(_, And(_, _, Equals(_, _, _, _)), And(_, eq @ Equals(_, _, Application(_, _), _), _)) => Some(eq)
+        case Implies(_, And(_, _, Top(_)), And(_, eq @ Equals(_, _, Application(_, _), _), _)) => Some(eq)
+        case eq @ Equals(_, _, Application(_, _), _) => Some(eq)
+        case _ => None
```

Comment:
How did you come up with this list of patterns? Could it be done as a visitor which accumulates a list of Rewrites and Equals instead?

Reply:
@ehildenb it cannot be done as a visitor that accumulates all rewrites and equals, because some instances of those patterns do not correspond to actual rules in the original semantics, and therefore should not be included in this list. This list of patterns corresponds more or less to the structure of the axioms generated from rules in ModuleToKore.java and was taken originally from the llvm backend.

Last I heard, the haskell backend was also doing something very similar in matching on the structure of axioms, but I'm not sure if that's still true. @ttuegel may know more.

Reply:
with that being said, I'm not opposed to moving this logic into the k frontend and sharing the same code path between this unit test and the llvm backend, if you'd like.

Reply:
If it's common logic, then I think it should be moved. This list of patterns still seems rather ad-hoc to me (not saying that it is ad-hoc, just that I can't seem to see how it's derived). But if it's common for all backends then it seems ok.